blog-content/the-complete-guide-to-opentelemetry-in-next-js.md

```diff
@@ -0,0 +1,394 @@
+  },
+};
+```
+
+
+### Enabling OpenTelemetry in Next.js
+
+Automatic Spans in Next.js
+
+Next.js has built-in OpenTelemetry spans for various parts of the application, including:
+- API routes (pages/api or app/api)
+- Page router (Pages Directory)
+- App router (App Directory)
+
+To enable verbose OpenTelemetry logging, set the following environment variable:
+
+```bash
+NEXT_OTEL_VERBOSE=1
+```
+
+With this enabled, Next.js will provide detailed span information in the logs, making it easier to debug and understand request flow.
+
+### Tracing API Requests and Page Loads
+
+#### Tracing API Routes
+
+Next.js automatically creates spans for API requests. For example, a request to /api/user will generate spans like:
+- next.js /api/user request
+- next.js /api/user handler
+- fetch /external-api (if the API calls another service)
+
+If you have an API route in pages/api/user.ts:
+
+```javascript
+export default async function handler(req, res) {
+  res.status(200).json({ name: 'John Doe' });
+}
+```
+
+You can observe spans for:
+✅ Incoming request (GET /api/user)
+✅ Response timing
+✅ Any internal or external API calls
+
+#### Tracing Page Router (Pages Directory)
+
+For traditional Next.js pages (pages/index.js):
+
+```javascript
+export default function HomePage() {
+  return <h1>Welcome to Next.js</h1>;
+}
+```
+
+When a user navigates to /, Next.js automatically creates spans for:
+- next.js / (page load)
+- next.js render /index
+- fetch spans (if the page makes API calls)
+
+#### Tracing App Router (App Directory)
+
+If you’re using the Next.js App Router (app/page.tsx), spans will include:
+- Page load spans (next.js /app/page.tsx)
+- Server component execution spans (next.js /layout.tsx)
+- Data fetching spans for fetch() calls in server components
+
+For example, an app router component:
+
+```javascript
+export default function Page() {
+  return <h1>Welcome to the App Router!</h1>;
+}
+```
+will automatically generate spans for rendering and hydration.
+
+### Logging in OpenTelemetry
+
+To correlate logs with traces, install OpenTelemetry logging:
+
+
+```bash
+yarn add @opentelemetry/sdk-logs
+```
+
+Modify `otel.js`:
+
+```javascript
+import { LoggerProvider } from '@opentelemetry/sdk-logs';
+import { ConsoleLogger } from '@opentelemetry/sdk-logs';
+
+const loggerProvider = new LoggerProvider();
+const logger = loggerProvider.getLogger('nextjs-logger');
+
+logger.emit({
+  severityText: 'INFO',
+  body: 'Application started',
+});
+```
+This allows you to log messages that correlate with traces.
+
+### Capturing Exceptions with Spans
+
+To capture exceptions inside traces, you can manually add error attributes to spans. Modify your API route:
+
+```javascript
+import { trace } from '@opentelemetry/api';
+
+export default async function handler(req, res) {
+  const span = trace.getActiveSpan();
+  
+  try {
+    throw new Error('Something went wrong!');
+  } catch (error) {
+    if (span) {
+      span.setAttribute('error', true);
```

Comment:
Nit, but is this attribute a convention we came up with? Wondering why it's included here. I assume it's to make it easier to find spans with exceptions?